core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,278 @@
+        private final String url;
+
+        public Redirection(String url) {
+            this.url = url;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            String url = Jenkins.get().getRootUrl() + this.url + restOfPath;
+            rsp.sendRedirect(302, url);
+        }
+    }
+
+    public String getRedirectUrl(Token token, String restOfPath) {
+        String resourceRootUrl = getResourceRootUrl();
+        if (!resourceRootUrl.endsWith("/")) {
+            resourceRootUrl += "/";
+        }
+        if (!restOfPath.startsWith("/")) {
+            // Unsure whether this can happen -- just be safe here
+            restOfPath = "/" + restOfPath;
+        }
+        return resourceRootUrl + getUrlName() + "/" + token.encode() + restOfPath;
+    }
+
+    private static String getResourceRootUrl() {
+        return ResourceDomainConfiguration.get().getUrl();
+    }
+
+    public Token getToken(DirectoryBrowserSupport dbs, StaplerRequest req) {
+        String dbsFile = req.getRestOfPath();
+
+        String completeUrl = req.getAncestors().get(0).getRestOfUrl();
+        completeUrl = completeUrl.substring(0, completeUrl.length() - dbsFile.length());
+
+        Authentication authentication = Jenkins.getAuthentication();
+        String authenticationName = authentication == Jenkins.ANONYMOUS ? "" : authentication.getName();
+
+        try {
+            return new Token(completeUrl, authenticationName, new Date().getTime());
+        } catch (Exception ex) {
+            LOGGER.log(Level.WARNING, "Failed to encode token for URL: " + completeUrl + " user: " + authenticationName, ex);
+        }
+        return null;
+    }
+
+    /**
+     * Implements the browsing support for a specific {@link DirectoryBrowserSupport} like permission check.
+     */
+    private static class InternalResourceRequest {
+        private final String authenticationName;
+        private final String browserUrl;
+
+        InternalResourceRequest(@Nonnull String browserUrl, String authenticationName) {
+            this.browserUrl = browserUrl;
+            this.authenticationName = authenticationName;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            AccessControlled requestRoot = Jenkins.get();
+            String requestUrlSuffix = this.browserUrl;
+
+            LOGGER.fine(() -> "Performing a request as authentication: " + authenticationName + " to object: " + requestRoot + " and restOfUrl: " + requestUrlSuffix + " and restOfPath: " + restOfPath);
+
+            Authentication auth = Jenkins.ANONYMOUS;
+            if (authenticationName != null) {
+                User user = User.getById(authenticationName, false);
+                if (user != null) {
+                    auth = user.impersonate();
```

Comment:
So this will verify that the user is still valid in the security realm.

Reply:
Good point, wrapping this is a try/catch to handle an exception here.